File: src/com/leo/leetcode/algorithm/q0000/Q77.java
package com.leo.leetcode.algorithm.q0000;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * 给定两个整数 n 和 k，返回 1 ... n 中所有可能的 k 个Q数的组合。
 * 链接：https://leetcode-cn.com/problems/combinations/
 */
public class Q77 {
    public static void main(String[] args) {
        System.out.println(new Q77().combine(4, 2)); // [[1, 2], [1, 3], [1, 4], [2, 3], [2, 4], [3, 4]]
    }

    public List<List<Integer>> combine(int n, int k) {
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) nums[i] = i + 1;
        List<List<Integer>> out = new ArrayList<>();
        build(out, null, nums, 0, k);
        return out;
    }

    void build(List<List<Integer>> out, LinkedList<Integer> ans, int[] nums, int offset, int k) {
        if (ans == null) ans = new LinkedList<>();
        for (int i = offset; i < nums.length; i++) {
            ans.add(nums[i]);
            if (k == 1) {
                out.add(new ArrayList<>(ans));
            } else {
                build(out, ans, nums, ++offset, k - 1);
            }
            ans.removeLast();
        }
    }
}

File: py/q0000/Q77.py
"""
 * 给定两个整数 n 和 k，返回范围 [1, n] 中所有可能的 k 个数的组合。
 * 你可以按 任何顺序 返回答案。
 * 提示：
 * 1、1 <= n <= 20
 * 2、1 <= k <= n
 * 链接：https://leetcode.cn/problems/combinations/
"""
from itertools import combinations
from typing import List


class Solution:

    def combine(self, n: int, k: int) -> List[List[int]]:
        return [list(l) for l in combinations([i for i in range(1, n + 1)], k)]


if __name__ == '__main__':
    # [[1]]
    print(Solution().combine(1, 1))
    # [[1, 2], [1, 3], [1, 4], [2, 3], [2, 4], [3, 4]]
    print(Solution().combine(4, 2))